login_server/src/cpp/controller/User.cpp

```diff
@@ -194,28 +194,38 @@ namespace controller {
 		return json;
 	}
 
-	int User::login(const std::string& password)
+	Poco::AutoPtr<SecretKeyCryptography> User::createSecretKey(const std::string& password)
 	{
-		if (!mPassword.isNull() && mPassword->hasKey()) {
-			return 2;
-		}
 		auto observer = SingletonTaskObserver::getInstance();
 
-		std::unique_lock<std::shared_mutex> _lock(mSharedMutex);
-		assert(mPassword.isNull());
-
 		auto model = getModel();
 		auto email_hash = observer->makeHash(model->getEmail());
 		if (observer->getTaskCount(email_hash, TASK_OBSERVER_PASSWORD_CREATION) > 0) {
-			return -3;
+			return nullptr;
 		}
 		observer->addTask(email_hash, TASK_OBSERVER_PASSWORD_CREATION);
 		Poco::AutoPtr<SecretKeyCryptography> authenticated_encryption(new SecretKeyCryptography);
 		assert(!authenticated_encryption.isNull() && model);
 		authenticated_encryption->createKey(model->getEmail(), password);
 
 		observer->removeTask(email_hash, TASK_OBSERVER_PASSWORD_CREATION);
+		return authenticated_encryption;
+	}
 
+	int User::login(const std::string& password)
+	{
+		std::unique_lock<std::shared_mutex> _lock(mSharedMutex);
+
+		if (!mPassword.isNull() && mPassword->hasKey()) {
+			return 2;
+		}
+		assert(mPassword.isNull());
```

Comment:
This looks very fishy. You cannot assert password to be null since it could be not Null and `hasKey = false`?!
Why do you assert that the users password is null on login?